### PR TITLE
Added `chain_err` method to Error implementations

### PR DESCRIPTION
**Pull request by TedDriggs**
Add a new method `Error::caused(self, error: F) -> Error where F : FnOnce() -> EK, EK : Into<ErrorKind>` to the generated error type. This enables chaining on error instances exposed by methods such as `Result::map_err` without inventing a new Result wrapper. This aims to address #40 

**Open Questions**
1. Should this method be named `chain_err` to align with `ResultExt`?
1. Does this change avoid the inference issues from #62? All tests pass, but I'm not sure if the tests covered that case.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Unreleased
 
+- Add a new method for `Error`: `caused`. _PR to be added_
```
Comment (Yamakaky) [inline CHANGELOG.md]: You can now add the PR number ^^

File: src/error_chain.rs [hunk 2]
```diff
@@ -145,6 +145,12 @@ macro_rules! error_chain_processed {
             pub fn backtrace(&self) -> Option<&$crate::Backtrace> {
                 self.1.backtrace()
             }
+            
+            /// Extends the error chain with a new entry.
+            pub fn caused<F, EK>(self, error: F) -> $error_name 
+                where F: FnOnce() -> EK, EK: Into<$error_kind_name> {
+                $error_name::with_chain(self, Self::from_kind(error().into()))
+            }
```
Comment (Yamakaky) [inline src/error_chain.rs]: Did you consider adding this to the `ChainedErr` trait?

Comment (TedDriggs) [inline src/error_chain.rs]: I had not; it looks like it should be straightforward enough to include. I'll put that in now.

**Comment by Yamakaky** [comment]
Sorry for the delay. Could you add some doc in lib.rs?

**Comment by TedDriggs** [comment]
Yes. Is `caused` the name you'd like me to use, or should I go with `chain_err` or something instead?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # Unreleased
 
+- [Add a new method for `Error`: `caused`.](https://github.com/brson/error-chain/pull/141)
```
Comment (kgv) [inline CHANGELOG.md]: Need to change description `caused` -> `chain_err`

Comment (TedDriggs) [inline CHANGELOG.md]: Good catch; I've fixed that and pushed a new update.

**Comment by TedDriggs** [comment]
@kgv / @Yamakaky, please let me know if you need anything else for this to move forward.

**Comment by Yamakaky** [comment]
Sorry, I had a lot of things to do these times. SGTM!

**Comment by TedDriggs** [comment]
@Yamakaky thanks for the merge; do you mind publishing a new release with this and the other changes sitting in master?

**Pull request merged**